### DISCUSSION TITLE
Tensor to Numpy style string

### DISCUSSION THREAD

**Discussion by dayo05** [xamarin/TorchSharp#491] [Show and tell]
I think who learns ML will friendly with python, numpy. But, torchsharp uses julia's method to convert tensor to string.
So I made Tensor to numpy string.
I don't think this is best way but this is able to be one of the method.

```cs
public static class Numpy
{
    public static string ToNumpyString(this Tensor t)
        => ToNumpyString(t, t.ndim);

    private static string ToNumpyString(Tensor t, long mdim, bool isFCreate = true)
    {
        var dim = t.dim();
        var sb = new StringBuilder(isFCreate ? string.Join("", Repeat(' ', (int)(mdim - dim))) : "");
        sb.Append('[');
        var currentSize = t.size()[0];
        if (dim == 1)
        {
            if (currentSize <= 6)
            {
                for (var i = 0; i < currentSize - 1; i++)
                {
                    sb.Append(t[i].ToSingle());
                    sb.Append(' ');
                }

                sb.Append(t[currentSize - 1].ToSingle());
            }
            else
            {
                for (var i = 0; i < 3; i++)
                {
                    sb.Append(t[i].ToSingle());
                    sb.Append(' ');
                }

                sb.Append(" ... ");
                
                for (var i = currentSize - 3; i < currentSize - 1; i++)
                {
                    sb.Append(t[i].ToSingle());
                    sb.Append(' ');
                }

                sb.Append(t[currentSize - 1].ToSingle());
            }
        }
        else
        {
            var newline = string.Join("", Repeat(Environment.NewLine, (int) dim - 1).ToList());
            if (currentSize <= 6)
            {
                sb.Append(ToNumpyString(t[0], mdim, false));
                sb.Append(newline);
                for (var i = 1; i < currentSize - 1; i++)
                {
                    sb.Append(ToNumpyString(t[i], mdim));
                    sb.Append(newline);
                }
                sb.Append(ToNumpyString(t[currentSize - 1], mdim));
            }
            else
            {
                sb.Append(ToNumpyString(t[0], mdim, false));
                sb.Append(newline);
                for (var i = 1; i < 3; i++)
                {
                    sb.Append(ToNumpyString(t[i], mdim));
                    sb.Append(newline);
                }

                sb.Append(string.Join("", Repeat(' ', (int) (mdim - dim))));
                sb.Append("...");
                sb.Append(newline);
                
                for (var i = currentSize - 3; i < currentSize - 1; i++)
                {
                    sb.Append(ToNumpyString(t[i], mdim));
                    sb.Append(newline);
                }
                sb.Append(ToNumpyString(t[currentSize - 1], mdim));
            }
        }

        sb.Append("]");
        return sb.ToString();
    }
}
```

**Comment by NiklasGustafsson** [comment]
@dayo05 -- I think it would be valuable to also add some of the arguments from the ToString() method, especially the format and cultureInfo args.

```C#
public string ToString(bool withData string fltFormat = "g5", int width = 100, CultureInfo? cultureInfo = null)
```

Also, wouldn't you want to format things depending on type? Your code looks like assumes everything is a 32-bit float. If it's useful, we could make 'Tensor.PrintValue' public.

**Comment by NiklasGustafsson** [comment]
Personally, I find NumPy tensor strings completely unintelligible for anything beyond 2 dimensions, but I realize that it's very subjective. It's cool to see you hack this up -- it could go in the `TensorExtensionMethods` class, I guess.

In that case, `PrintValue()` wouldn't have to be public.

**Reply by NiklasGustafsson** [reply]
@dayo05 -- did you want to put this in a PR? I think it belongs in `TensorExtensionMethods`, by the way.

**Reply by NiklasGustafsson** [reply]
@dayo05 -- just a quick reminder: I think this is great as an added feature do you want to put it in a PR?

**Reply by dayo05** [reply]
@NiklasGustafsson I'm working on this and I think making this as default ToString method and let privious way as `ToString(false)`. Because, almost of docs are explains about numpy style and this will be better for beginners

**Reply by NiklasGustafsson** [reply]
@dayo05 -- ToString(false) won't work, since that's meant to mean reverting to the data-free 'ToString()' (only element type, shape, device).

I don't think I agree with changing the default -- it's important to adhere to the PyTorch APIs, but I'm not convinced that it necessary to have diagnostic output be identical. I personally find the PyTorch str() representation of tensors completely unreadable once you get above two dimensions, while the (Julia-inspired) TorchSharp make the dimensionality clear.

So, I like the idea of adding numpy-style str(), but not making it the default.

**Comment by GeorgeS2019** [comment]
[Related discussion](https://github.com/dotnet/TorchSharp/pull/680#issuecomment-1205657199)

**Reply by dayo05** [reply]
See also: #572

**Comment by NiklasGustafsson** [comment]
The numpy style of tensor strings has been integrated a while ago.
